Object oriented, PEP8 compliance and python3.5 support

XayOn opened the pull request:
I needed this to work with python3.5, I find this project a much better solution than those requiring GI installed (ugh) with loads of un-pythonic C-bindings-related lines of code. 

Along the way I ended up practically rewriting the project in the lastest three hours ( couldn't sleep...)

I hope you like it, I added
- Python up to 3.5 support
- OOP
- Better practices
- Separated tests for windows / linux on the doctests
- Using pbr for package and version management

BTW, I'd recommend adding this to pypi, wich would require a new name for the project, as zeroconf is already taken

XayOn committed:
Sorry, type is a reserved word, replacing it for type_

XayOn committed:
Almost complete rewrite of the project using:

- OOP
- pbr for better package handling
- Python >= 2.7 < 3.5 support
- Better practices
- PEP8
- Separate modules for windows and linux
- Independent doctests for both OS

XayOn committed:
Added support for python3.5 to readme

XayOn committed:
Don't encourage users to install system-wide packages with sudo.
They should know how to handle package installation (if they do
it with sudo, let it not be this package's fault ;-) )

boisgera commented:
Hi David,

Thanks for your efforts; obviously this project could use more love :)

There are many things in your contributions: some good/uncontroversial things,
some things I would need a bit of work to review / discuss / make up my mind, 
and there a few parts I don't like ;)

We can discuss this we would probably come up (eventually ;) with something that could
please both of us ... If that's ok I'll start with some basic issues so that we can handle 
these things separately.

Misc. comments (to be dispatched in issues):
- sudo or not sudo: I believe that there are good uses case for system-wide installs 
  and user installs; I don't mind also mentioning the user install (or even only this one), 
  but it's not only a matter of removing "sudo", the --user options should be also added.
- sh dependency: you are right, I should not ask the user to install it manually,
  it should be done automatically by the install (and maybe a pip command instead
  of the manual call to setup.py).
- type vs type_: despite type being a keyword, you can use it as a keyword argument,
  and it's arguably simpler for the end user; if I ever need the type function, I can still
  import it from the **builtin** module, so I wouldn't change this.
- py.test: never felt that I needed something beyond doctest and I still don't feel the need here,
  I am not sold :) 
  OK though to document a way to actually run the tests.
- pbr: I don't typically use it. I agree that the current setup.py is a mess and 
  that it's an improvement (nowadays, I use "about" to do something similar,
  but it's a pet project, pbr is more "standard").
  However, I lose the metadata that was in the zeroconf module, right?
- where are my ditz issues now ?
- the linux/window split: sure. The doctest split? Not so sure. Ideally the tests
  should be the same, so do we not having a duplication here?
  (I have not looked in details into your code on this point)
- python 2.7/3.5: I don't know specifically what you did to support 3.5, 
  can hardly comment. In this case, the ideal thing would be to setup 
  a Travis CI to test against both interpreters.
- pypi: yeah zeroconf is taken, this is a mess :).
  I'd like the module to be named zeroconf, but also the module to have
  the same name than the project. Maybe using a namespace (such as
  e.g. network.zeroconf) would be ok.

Cheers,

SB

XayOn commented:
> > > sudo or not sudo: I believe that there are good uses case for system-wide installs 
> > > and user installs; I don't mind also mentioning the user install (or even only this one), 
> > > but it's not only a matter of removing "sudo", the --user options should be also added.

Yep, installing documentation could be improved.

> > > sh dependency: you are right, I should not ask the user to install it manually,
> > > t should be done automatically by the install (and maybe a pip command instead
> > > of the manual call to setup.py).

I tought I added the pip install to the readme.

> > > type vs type_: despite type being a keyword, you can use it as a keyword argument,
> > > and it's arguably simpler for the end user; if I ever need the type function, I can still
> > > import it from the builtin module, so I wouldn't change this.

Kind of ugly, but it's your project, I agree on the simplicity for the user and as it is scope is the ZeroConf object it should not break anything

> > > py.test: never felt that I needed something beyond doctest and I still don't feel the need here,
> > > I am not sold :) 
> > > OK though to document a way to actually run the tests.

Yep, actually it's only there to document a way to run the tests without adding doctest call directly to the code.
Separate tests from code (yet it's still ok to have them in the doc)
Also, it makes a clearer way to integrate this with travis.

> > > pbr: I don't typically use it. I agree that the current setup.py is a mess and 
> > > that it's an improvement (nowadays, I use "about" to do something similar,
> > > but it's a pet project, pbr is more "standard").
> > > However, I lose the metadata that was in the zeroconf module, right?

I actually deleted it myself, probably. I ended up rewriting from scratch and probably forgot to copy them =)

> > > where are my ditz issues now ?

I'd totatlly import them on github or somewhere you don't actually need to have them with your code. But it's your project, you can discard that part of the merge (is a specific commit, made it that way to make my life simpler when remaking the package and so on, but it's a separate commit so you can revert it (actually, I was going to discard it, but It seems like I forgot about it)

> > > the linux/window split: sure. The doctest split? Not so sure. Ideally the tests
> > > should be the same, so do we not having a duplication here?
> > > (I have not looked in details into your code on this point)

Probably. But I see a real problem on using doctest on windows: it's harder to mock. And on windows, if you want travis tests, you HAVE to mock them.
I was thinking on improving those tests with py.test and Mocks (If you know a good way to make mocks (emulate mdns behavior, for example) I could do that, but again, I've only made that with py.test)

> > > python 2.7/3.5: I don't know specifically what you did to support 3.5, 
> > > can hardly comment. In this case, the ideal thing would be to setup 
> > > a Travis CI to test against both interpreters.

Totally, I made the tests locally with 2.7, 3.4 and 3.5, but travis should be the way to go, that's exactly what I tought.
The thing is, it's your project, you would need to create the travis project so the URL is the main project's url and not some forks' url like mine =)

> > > pypi: yeah zeroconf is taken, this is a mess :).
> > > I'd like the module to be named zeroconf, but also the module to have
> > > the same name than the project. Maybe using a namespace (such as
> > > e.g. network.zeroconf) would be ok.

I think the fact that this uses avahi and mdns tools and it's kinda-multiplatform but requires external programs should be somehow in the name. But I cannot find a real way to do that.

XayOn commented:
It seems I was wrong, I put the ditz deletion on another commit, I'll revert that part right away.

XayOn commented:
Oh, now I see why I separated doctests like that, I had three options:
- Having doctests in a separate function (wich is a really bad practice, doctests are to be made as near as possible to their tested code, that's why they're beautiful! (and they can be used in documentations, wich is cool too)
- Having doctests only in the linux one, as the windows one would break on linux, so it'll break on travis.
- Having doctests in both, and autodetect the OS before running them

The lastest part seems to be the best, but it would require either pytest or some hack like I did.
I'd really prefer pytest, but it seems you don't like it. Anyway, with pytest we could mock windows objects and have windows code really tested.

If you can find another elegant solution to adding continuous integration on this code ...

XayOn committed:
Recovered bug files

XayOn committed:
Back using reserved word type as argument, as requested by boisgera

XayOn committed:
Added more install options to the readme

XayOn committed:
Back using reserved word type as argument, as requested by boisgera

XayOn commented:
I updated this pull request with some of your suggestions.

XayOn committed:
Added travis and tox integration and python setup.py test

XayOn committed:
As I'm ignoring the windows one in pytest config (sorry, that would have to be run manually on windows machines, this distinction is not necesary anymore

XayOn committed:
Added comment about running tests on windows and using python setup.py test

XayOn committed:
Fixed indentation problems on linux test

XayOn commented:
The pending things are:

> py.test: never felt that I needed something beyond doctest and I still don't feel the need here,
> I am not sold :) 
> OK though to document a way to actually run the tests.
> the linux/window split: sure. The doctest split? Not so sure. Ideally the tests
> should be the same, so do we not having a duplication here?
> (I have not looked in details into your code on this point)

Testing, in general. I really think it'd need pytest to make cross-platform mocked tests.

> python 2.7/3.5: I don't know specifically what you did to support 3.5, 
> can hardly comment. In this case, the ideal thing would be to setup 
> a Travis CI to test against both interpreters.

I added a travis.yml file, and made "python setup.py test" to work with doctest, configured it to ignore windows file so we can automate it and added an entry in readme about windows testing.

> pypi: yeah zeroconf is taken, this is a mess :).
> I'd like the module to be named zeroconf, but also the module to have
> the same name than the project. Maybe using a namespace (such as
> e.g. network.zeroconf) would be ok.

I said earlier that I tought it should reflect the way this is implemented (with system calls), but I'm starting to not-like-that-much the idea and like yours way better.
Anyway, I think that's beyond the scope of this PR, and should be addressed separately =)

XayOn committed:
Ups, copypasted tox.ini from another file and forgot to change project name, sorry

XayOn committed:
Add note on dep instalation

XayOn committed:
It's avahi-utils, not avahi-tools

XayOn committed:
This does not build in travis, it seems like some strange conflict with pbr versioning, also, I'm adding a static version string on setup.cfg, as boisguera commented, I lost it when migrating to pbr from the metadata-based system

XayOn committed:
Seems like pbr is having conflict with versions

XayOn committed:
Added a vagrantfile to test locally on clean environments

XayOn committed:
Forcing setuptools to a bigger version

XayOn committed:
Added sh as a package requirement

XayOn commented:
Any news?
